web/src/message_events.js

```diff
@@ -101,6 +104,13 @@ export function insert_new_messages(messages, sent_by_this_client, deliver_local
         unread_ui.update_unread_counts();
     }
 
+    // If it is a local echo, we try to patch the messages to the
+    // data structures in `echo_state` before we update the stream
+    // sidebar.
+    if (deliver_locally) {
+        messages.map((message) => patch_local_message(message));
+    }
+
```

Comment:
This is important because previously `echo_state.set_message_waiting_for_id` and `echo_state.set_message_waiting_for_ack` were called after updating stream sidebar. 

This used to work fine previously since there we directly updated the max ids in `stream_topic_history` with `local_id`. However, over here, since we are dependent on the `echo_state` for topics ordering, we need to update it before stream sidebar is updated.

Reply:
OK, this strategy makes sense but I may need to do some tweaks to the comments for posterity.